File: leetcode/2001-3000/2101-2200/2121-2130/2127-maximum-employees-to-be-invited-to-a-meeting/MaximumEmployeesToBeInvitedToAMeeting.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class MaximumEmployeesToBeInvitedToAMeeting {
  public static void main(String[] args) {
    System.out
        .println(new Solution().maximumInvitations(new int[] { 1, 0, 3, 2, 5, 6, 7, 4, 9, 8, 11, 10, 11, 12, 10 }));
  }
}

class Solution {
  int maxCircleSize(int[] favourites) {
    int n = favourites.length;
    int max = 0;
    boolean[] seen = new boolean[n];

    for (int i = 0; i < n; i++) {
      int current = i;

      if (seen[current]) {
        continue;
      }

      List<Integer> circle = new ArrayList<>();

      while (!seen[current]) {
        seen[current] = true;
        circle.add(current);
        current = favourites[current];
      }

      for (int j = 0; j < circle.size(); j++) {
        if (circle.get(j) == current) {
          max = Math.max(max, circle.size() - j);
          break;
        }
      }
    }

    return max;
  }

  int maxChainLength(int[] favourites) {
    int n = favourites.length;
    int[] levels = new int[n];
    int[] distances = new int[n];
    Arrays.fill(distances, 1);

    for (int f : favourites) {
      levels[f]++;
    }

    Queue<Integer> queue = new LinkedList<>();

    for (int i = 0; i < n; i++) {
      if (levels[i] == 0) {
        queue.offer(i);
      }
    }

    while (!queue.isEmpty()) {
      int current = queue.poll();

      int next = favourites[current];

      distances[next] = Math.max(distances[current] + 1, distances[next]);

      levels[next]--;

      if (levels[next] == 0) {
        queue.offer(next);
      }
    }

    int total = 0;

    for (int i = 0; i < n; i++) {
      if (i == favourites[favourites[i]]) {
        total += distances[i];
      }
    }

    return total;
  }

  public int maximumInvitations(int[] favourites) {
    return Math.max(maxCircleSize(favourites), maxChainLength(favourites));
  }
}

File: leetcode/2001-3000/2101-2200/2121-2130/2127-maximum-employees-to-be-invited-to-a-meeting/MaximumEmployeesToBeInvitedToAMeeting.ts
const maxCircleSize = (favourites: number[]): number => {
  let size = 0;
  let max = 1;

  const seens: Set<number> = new Set();

  for (let i = 0; i < favourites.length; i++) {
    const currentCircle: Set<number> = new Set();
    size = 0;
    let node = i;

    while (!currentCircle.has(node) && !seens.has(node)) {
      currentCircle.add(node);
      size++;
      node = favourites[node];
    }

    if (node === i) {
      max = Math.max(max, size);

      currentCircle.forEach(e => seens.add(e));
    }
  }

  return max;
}

const maxChainLength = (favourites: number[]): number => {
  const n = favourites.length;
  const levels: number[] = new Array(n).fill(0);
  const distances: number[] = new Array(n).fill(1);

  for (let i = 0; i < n; i++) {
    levels[favourites[i]]++;
  }

  const queue: number[] = [];

  for (let i = 0; i < n; i++) {
    if (levels[i] === 0) {
      queue.push(i);
    }
  }

  while (queue.length > 0) {
    const node = queue.shift()!;
    const next: number = favourites[node];

    distances[next] = Math.max(distances[node] + 1, distances[next]);
    levels[next]--;

    if (levels[next] === 0) {
      queue.push(next);
    }
  }

  return distances.filter((_, i) => i === favourites[favourites[i]])
    .reduce((accumulated, value) => accumulated + value, 0);
}

function maximumInvitations(favourites: number[]): number {
  return Math.max(maxCircleSize(favourites), maxChainLength(favourites));
};

console.log(maximumInvitations([1, 0, 3, 2, 5, 6, 7, 4, 9, 8, 11, 10, 11, 12, 10]));
